### DISCUSSION TITLE
Applying force boundary condition to a particular set of nodes

### DISCUSSION THREAD

**Discussion by chakra34** [idaholab/moose#23909] [Q&A Modules: General] [answered]
Hi,

I am very new to MOOSE, and currently am trying to compare ANSYS and MOOSE (the idea is to use more of MOOSE in future). The problem I am trying now is a 4-point beam bending test, where in ANSYS we put force boundary condition on a selected curve (set of nodes) as shown in the image below:
![image](https://user-images.githubusercontent.com/22983992/228921703-34b015f2-82d9-442a-b887-deaaa4fd0d87.png)

Now to do the same in MOOSE, I create the geometry in Cubit and make a nodeset at the same location (since sidesets are not supported in exodus format from cubit :( ), and I want to impose a NeumannBC or PressureBC at these nodes such that I can impose a force. But I see that MOOSE doesn't support applying such boundary condition to nodesets and expect sidesets.

Is there any possible solution or a kernel I can use to impose such BC in MOOSE ?

Thanks,
Aritra

**Comment by GiudGiud** [comment]
Hello

You may use this boolean in the [Mesh] block to convert nodesets to sidesets:
construct_side_list_from_node_list = true

Guillaume

**Reply by chakra34** [reply]
Yeah, I tried that but still getting the error

So this is my input file:
```
[Mesh]
  construct_side_list_from_node_list = true
  file=beam_geometry_for_moose_fine.e
  
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]
[Variables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
  [./disp_z]
  [../]
  [temp]
    initial_condition = 1523.0
  []
[]
[Kernels]
  [heat]
    type = Diffusion
    variable = temp
  []
[]
[Modules/TensorMechanics/Master/All]
  strain = FINITE
  incremental = true
  add_variables = true
  generate_output = 'stress_yy elastic_strain_yy'
  use_automatic_differentiation = false
[]
[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1.732e9
    poissons_ratio = 0.262
    block = 1
  []
[]

[Materials]
  [./elastic_strain]
    type = ComputeMultipleInelasticStress
    block = 1
    inelastic_models = 'creep'
    tangent_operator = nonlinear
  [../]
  [./creep]
    type = PowerLawCreepStressUpdate
    block = 1
    coefficient = 2.43e-24
    n_exponent = 1
    activation_energy = 0.0
    absolute_tolerance = 1e-11
    relative_tolerance = 1e-11
  [../]
[]

[AuxVariables]
  [./aux_creep_strain]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./vM_elastic_strain]
    order = CONSTANT
    family = MONOMIAL
  [../]  
  [./von_mises_stress]
	#Dependent variable used to visualize the Von Mises stress
    order = CONSTANT
    family = MONOMIAL
  [../]

[]

[AuxKernels]
  [./creep_aux]
    type = MaterialRealAux
    variable = aux_creep_strain
    property = effective_creep_strain
    execute_on = timestep_end
  [../]
  [./elastic_strain_aux]
    type = RankTwoScalarAux
    variable = vM_elastic_strain
    rank_two_tensor = elastic_strain
    scalar_type = VonMisesStress
    execute_on = timestep_end
  [../]
  [./von_mises_kernel_stress]
   #Calculates the von mises stress and assigns it to von_mises_stress
    type = RankTwoScalarAux
    variable = von_mises_stress
    rank_two_tensor = stress
    execute_on = timestep_end
    scalar_type = VonMisesStress
  [../]
[]

[Functions]
  [top_pull]
    type = PiecewiseLinear
    x = '0 1'
    y = '1 1'
  []
[]

[BCs]
  [u_x_zero]
    type = DirichletBC
    variable = disp_x
    boundary = 'u_x_zero'
    value = 0
  [../]
  [u_y_zero]
    type = DirichletBC
    variable = disp_y
    boundary = 'u_y_zero_left u_y_zero_right'
    value = 0
  [../]
  [u_z_zero]
    type = DirichletBC
    variable = disp_z
    boundary = 'u_z_zero'
    value = 0.0
  [../]
  [top_load_pressure]
    type = Pressure
    variable = disp_y
    boundary = 'force_y_left force_y_right'
    factor = -10.0e6
    function = top_pull
  []
  
[]



[Postprocessors]
  [./eps_crp_mag]
    type = ElementAverageValue
    variable = aux_creep_strain
  [../]
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient

  solve_type = 'Newton'

  start_time = 0.0
  dt    = 1.0
  dtmax = 2.0
  dtmin = 0.1
  nl_abs_tol = 1e-6
  end_time = 1
 []

[Outputs]
  print_linear_residuals = true
  perf_graph = true
  csv = true
  [./out]
    type = Exodus
    elemental_as_nodal = true
  [../]
  [./console]
    type = Console
    max_rows = 25
  [../]
[]
```
And this is the error:
```
*** ERROR ***
/lstr/home/chakraboa/Moose_runs/StaticFatigue/power_law_creep_plasticity_beam/creep_simulations/creep_with_force_bc.i:140: (BCs/top_load_pressure/boundary):
    the following side sets (ids) do not exist on the mesh: force_y_right (5), force_y_left (6)
    
    MOOSE distinguishes between "node sets" and "side sets" depending on whether 
    you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding 
    to your side sets are constructed for you by default.
    
    Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
    Note: If you are running with adaptivity you should prefer using side sets.
````
force_y_left and force_y_right are node sets that I made in cubit.
-Aritra

**Reply by GiudGiud** [reply]
can you use a FileMeshGenerator instead of the Mesh/file syntax?

https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html

**Reply by chakra34** [reply]
I tried the following:
```
[Mesh]
 [fmg]
  type = FileMeshGenerator
  file=beam_geometry_for_moose_fine.e
  construct_side_list_from_node_list = true
 [] 
[]
```
And I get the same error.

I was also trying to create sidesets using SideSetsFromPointsGenerator,
```
[Mesh]
 [fmg]
  type = FileMeshGenerator
  file=beam_geometry_for_moose_fine.e
 []

  [./sidesets_right]
    type = SideSetsFromPointsGenerator
    input = fmg
    points = '0.01   0.0015  0.004
              0.01   0.0015  0.003
              0.01   0.0015  0.002
              0.01   0.0015  0.001
              0.01   0.0015  0.000
              0.01   0.0015 -0.001
              0.01   0.0015 -0.002
              0.01   0.0015 -0.003
              0.01   0.0015 -0.004'
    new_boundary = 'side_force_y_right'
  []

  [./sidesets_left]
    type = SideSetsFromPointsGenerator
    input = fmg
    points = '-0.01   0.0015  0.004
              -0.01   0.0015  0.003
              -0.01   0.0015  0.002
              -0.01   0.0015  0.001
              -0.01   0.0015  0.000
              -0.01   0.0015 -0.001
              -0.01   0.0015 -0.002
              -0.01   0.0015 -0.003
              -0.01   0.0015 -0.004'
    new_boundary = 'side_force_y_left'
  []
 
[]

```
But that also gave the following error:
*** ERROR ***
The following error occurred in the object "sidesets_right", of type "SideSetsFromPointsGenerator".

point list and boundary list are not the same length

-Aritra

**Reply by GiudGiud** [reply]
It should be this
```
[Mesh]
  construct_side_list_from_node_list = true
  [fmg]
  type = FileMeshGenerator
  file=beam_geometry_for_moose_fine.e
  []
[]

**Reply by GiudGiud** [reply]
I've never used SidesetsFromPointGenerator but the docs there seem to say you just need to specify one point and it tries to paint around it
https://mooseframework.inl.gov/source/meshgenerators/SideSetsFromPointsGenerator.html

**Reply by chakra34** [reply]
Still got the error with:
[Mesh]
  construct_side_list_from_node_list = true
  [fmg]
  type = FileMeshGenerator
  file=beam_geometry_for_moose_fine.e
  []
[]

that sidesets do not exist...



Hmm..ohh okay. I was trying to put the nodal coordinates that would make the sideset with these coordinates, my bad.

**Reply by GiudGiud** [reply]
can you try generating the mesh with --mesh-only on the command line

It will give you a summary of the execution

**Reply by hugary1995** [reply, marked as answer]
@chakra34 I see that you have a 3D mesh, do the nodes belong to a line or a surface?

I am asking because I am not sure if `construct_side_list_from_node_list` is smart enough to handle the "line" case. Our convention is that side set of a 3D mesh is a collection of element _faces_ (as in topologically 2D faces).

**Reply by chakra34** [reply]
@GiudGiud 
This is the solution for mesh-only run:
```
 Mesh Information:
  elem_dimensions()={3}
  spatial_dimension()=3
  n_nodes()=4545
    n_local_nodes()=610
  n_elem()=3200
    n_local_elem()=400
    n_active_elem()=3200
  n_subdomains()=1
  n_partitions()=8
  n_processors()=8
  n_threads()=1
  processor_id()=0
  is_prepared()=true
  is_replicated()=true

 Mesh Bounding Box:
  Minimum: (x,y,z)=( -0.0508,  -0.0015,   -0.004)
  Maximum: (x,y,z)=(  0.0508,   0.0015,    0.004)
  Delta:   (x,y,z)=(  0.1016,    0.003,    0.008)

 Mesh Element Type(s):
  HEX8

 Mesh Nodesets:
  Nodeset 1 (u_x_zero), 9 nodes
   Bounding box minimum: (x,y,z)=(       0,  -0.0015,   -0.004)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 2 (u_z_zero), 2 nodes
   Bounding box minimum: (x,y,z)=( -0.0508,  -0.0015,    0.004)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 3 (u_y_zero_left), 9 nodes
   Bounding box minimum: (x,y,z)=(  -0.032,  -0.0015,   -0.004)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 4 (u_y_zero_right), 9 nodes
   Bounding box minimum: (x,y,z)=(   0.032,  -0.0015,   -0.004)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 5 (force_y_right), 9 nodes
   Bounding box minimum: (x,y,z)=(    0.01,   0.0015,   -0.004)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 6 (force_y_left), 9 nodes
   Bounding box minimum: (x,y,z)=(   -0.01,   0.0015,   -0.004)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)

 Mesh Sidesets:
  None

 Mesh Edgesets:
  None

 Mesh Subdomains:
  Subdomain 1 (block_1): 3200 elems (HEX8, 3200 active), 4545 active nodes
   Volume: 2.4384e-06
   Bounding box minimum: (x,y,z)=( -0.0508,  -0.0015,   -0.004)
   Bounding box maximum: (x,y,z)=(  0.0508,   0.0015,    0.004)
   Bounding box delta: (x,y,z)=(  0.1016,    0.003,    0.008)
  Global mesh volume = 2.4384e-06
```


=========================

@hugary1995 
They do not belong to a specific face boundary, I created an "arbitrary" curve when I generated the mesh in Cubit and assigned the nodes on that curve as the particular nodesets. So yeah, they are not a collection of the element faces. The geometry (and the curve) is the same as I have put in my screen shot, and the red line is where I want to apply the force boundary condition (basically the nodes that are on the red edge).

-Aritra

**Reply by hugary1995** [reply]
In that case I think you should use DiracKernels since you can't create a sideset out of it.

**Reply by hugary1995** [reply]
or maybe there is some way to do integration on "edgesets" instead of "sidesets" that I am not aware of...

**Reply by GiudGiud** [reply]
Could DiracKernels on nodesets work?

**Reply by hugary1995** [reply]
No, but you can simply provide the coordinates of the two nodes. @chakra34 will have to do the hand calculation to divide the total force by the number of points.

**Comment by lynnmunday** [comment]
You can make a sideset in cubit that will be exported to the exodus mesh:
```reset
create surface rectangle width 5 height 3 zplane
mesh surface all
block 1 surface all
sideset 101 add curve 2
sideset 101 name 'centerline'
export mesh "rectangle.e"  overwrite 
```

**Reply by chakra34** [reply]
@lynnmunday thanks a lot for the reply. I actually had tried it but it did not work. When I enquired with Cubit customer support they mentioned:

"Following up on my last email, we've confirmed that the Exodus format does not support sidesets comprised of curves (element edges) on 3D meshes.  You can find this described in the Exodus documentation here: https://sandialabs.github.io/seacas-docs/exodusII-new.pdf in section 4.11"

So for example when I add the sideset in Cubit it shows up:

![image](https://user-images.githubusercontent.com/22983992/228948601-b517e6db-53f6-421a-ad9f-651a3ea115bf.png)




But when I export the sideset do not transfer to exodus:


![image](https://user-images.githubusercontent.com/22983992/228948728-18e3393e-e321-4676-953b-0d844c2d14dd.png)

-Aritra

**Reply by hugary1995** [reply]
@lynnmunday your script works because it's creating a curve sideset on a 2D surface mesh.

**Reply by chakra34** [reply]
@GiudGiud @hugary1995 @lynnmunday 

btw, I was able to get a work around for now using the NodalKernel with type: "NodalGravity" which I guess you can use to apply some gravity force on a selected nodes. So I have the following block:

[NodalKernels]
  [./force_y2]
    type = NodalGravity
    variable = disp_y
    boundary = 'force_y_left force_y_right'
    gravity_value = 1.0 # inverse of nodal mass at cantilever end
    function = force
    mass = 1.0 
  [../]
[]

[Functions]
  [./force]
    type = PiecewiseLinear
    x = '0.0  1.0 '
    y = '-3.3 -3.3'
  [../]
[]

But I am a bit confused in the way the nodal gravity force is calculated. Is it like f = mg?

Looking at the source code I see it returns :

-mass * _gravity_value * _function.value(_t + _alpha * _dt, (*_current_node))

So in my case, does that mean that the **nodes in the selected boundary** (i.e., my nodesets) will have a force of 3.3 N or is it something completely different than what I am thinking it to be ?

Again, I really appreciate for your time and help.

-aritra

**Reply by lynnmunday** [reply]
I haven't used this object but what you are describing seems correct based on the code.  You can use the tag system to output the residual from the gravity kernel to see what the force is, like in this test to make sure its correct:
moose/modules/tensor_mechanics/test/tests/torque_reaction/torque_reaction_3D.i
Then you should be able to look at the saved in lagrange aux variable you create and they should be 3.3 like you expect.

**Reply by chakra34** [reply]
Thanks a lot @lynnmunday ! I will look into it. When you say the tag system, do you have a documentation ? I am sorry I am very new to MOOSE (basically since two days :( )

-Aritra

**Reply by lynnmunday** [reply]
I couldn't find where extra_tag_vectors is documented. The auxvariable for outputting them is here:
https://mooseframework.inl.gov/source/auxkernels/TagVectorAux.html
I don't know how they work but it seems like they are "tagging" certain contributions to the residual into a vector so that they can be output with that nodal auxvariable.

**Reply by chakra34** [reply]
got it, I will look into it then.

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/source/interfaces/TaggingInterface.html has some docs on tagging, not a lot
